GowerMetric.py

```diff
@@ -0,0 +1,371 @@
+            or self.dtypes[i] == DataType.BINARY_SYMMETRIC
+            else False
+            for i in range(len(self.dtypes))
+        ]
+        self._ratio_scale_idx = [
+            True if self.dtypes[i] == DataType.RATIO_SCALE else False
+            for i in range(len(self.dtypes))
+        ]
+
+        # Sums of columns of given type
+        self._cat_nom_num = sum(self._cat_nom_idx)
+        self._ratio_scale_num = sum(self._ratio_scale_idx)
+
+    def __call__(
+        self,
+        vector_1: np.ndarray,
+        vector_2: np.ndarray,
+    ):
+        # TODO - no check for fit
+        # sklearn.utils.validation.check_is_fitted(self)
+        if self.n_features_in_ != len(vector_1) | len(vector_2):
+            assert len(vector_1) == len(vector_2)
+
+        # Distance for cat_nom and bin_sym columns
+        if self._cat_nom_num:
+            cat_nom_cols_1 = vector_1[self._cat_nom_idx]
+            cat_nom_cols_2 = vector_2[self._cat_nom_idx]
+            cat_nom_dist = 1.0 - (cat_nom_cols_1 == cat_nom_cols_2)
+
+            # Weighted variant - dot product with according self.weights columns
+            # Unweighted - a sum of distances of between all columns
+            if self.weights is not None:
+                cat_nom_dist = cat_nom_dist @ self.weights[self._cat_nom_idx]
+            else:
+                cat_nom_dist = cat_nom_dist.sum()
+        else:
+            cat_nom_dist = 0.0
+
+        # Distance for ratio_scale columns
+        if self._ratio_scale_num:
+            ratio_scale_cols_1 = vector_1[self._ratio_scale_idx]
+            ratio_scale_cols_2 = vector_2[self._ratio_scale_idx]
+            ratio_dist = np.abs(ratio_scale_cols_1 - ratio_scale_cols_2)
+
+            ones = ratio_dist >= self.ranges_[self._ratio_scale_idx]
+            zeros = ratio_dist <= self.h_[self._ratio_scale_idx]
+
+            if ratio_scale_cols_1.size == 1:
+                if ones:
+                    ratio_dist = 1.0
+                elif zeros:
+                    ratio_dist = 0.0
+                else:
+                    ratio_dist /= self.ranges_[self._ratio_scale_idx]
+            else:
+                ratio_dist /= self.ranges_[self._ratio_scale_idx]
+                ratio_dist[zeros] = 0.0
+                ratio_dist[ones] = 1.0
+
+            if self.weights is not None:
+                ratio_dist = ratio_dist @ self.weights[self._ratio_scale_idx]
+            else:
+                ratio_dist = ratio_dist.sum()
+        else:
+            ratio_dist = 0.0
+
+        # Sum of all distances
+        distance = cat_nom_dist + ratio_dist
+
+        # Normalization
+        distance /= self.n_features_in_
+
+        return distance
+
+    def fit(self, X):
+        self.ranges_ = np.zeros(self.dtypes.size)
+        self.h_ = np.zeros(self.dtypes.size)
+        self.n_features_in_ = self.dtypes.size
+
+        for i in range(self.dtypes.size):
+            if self.dtypes[i] == DataType.RATIO_SCALE:
```

Comment:
Can this be vectorized? Now you process each variable separately, which is not terribly important, since `.fit()` is called only once, but still pretty important. You could do:
`X_ratio_scale = X[:, self.dtypes == DataType.RATIO_SCALE]`

And then work on that matrix. I am not 100% sure if all functions will work that way (e.g. `np.percentile()`), but they should.